qdrant_client/local/local_collection.py

```diff
@@ -1387,6 +1388,112 @@ def recommend_groups(
+        offsets_col = []
+
+        offset_by_id = {point_id: idx for idx, point_id in enumerate(ids)}
+
+        for row_offset, scored_points in enumerate(all_scores):
+            for scored_point in scored_points:
+                offsets_row.append(row_offset)
+                offsets_col.append(offset_by_id[scored_point.id])
+
+        # flatten the scores
+        scores = []
+        for sample_scores in all_scores:
+            for score in sample_scores:
+                scores.append(score.score)
+
+        return types.SearchMatrixOffsetsResponse(
+            offsets_row=offsets_row,
+            offsets_col=offsets_col,
+            scores=scores,
+            ids=ids,
+        )
+
+    def search_distance_matrix_pairs(
+        self,
+        query_filter: Optional[types.Filter] = None,
+        limit: int = 3,
+        sample: int = 10,
+        using: Optional[str] = None,
+    ) -> types.SearchMatrixPairsResponse:
+        ids, all_scores = self._search_distance_matrix(
+            query_filter=query_filter, limit=limit, sample=sample, using=using
+        )
+        pairs = []
+        for sample_id, sample_scores in list(zip(ids, all_scores)):
+            for sample_score in sample_scores:
+                pairs.append(
+                   types.SearchMatrixPair(
+                        a=sample_id,
+                        b=sample_score.id,
+                        score=sample_score.score
+                    )
+                )
+
+        return types.SearchMatrixPairsResponse(
+            pairs=pairs,
+        )
+
+    def _search_distance_matrix(
+        self,
+        query_filter: Optional[types.Filter] = None,
+        limit: int = 3,
+        sample: int = 10,
+        using: Optional[str] = None,
+    ) -> Tuple[List[ExtendedPointId], List[List[ScoredPoint]]]:
+        samples: List[ScoredPoint] = []
+        search_in_vector_name = using if using is not None else DEFAULT_VECTOR_NAME
+        # Sample random points from the whole collection to filter out the ones without vectors
+        # TODO: use search_filter once with have an HasVector like condition
+        candidates = self._sample_randomly(len(self.ids), query_filter, False, search_in_vector_name)
+        for candidate in candidates:
+            # check if enough samples are collected
+            if len(samples) == sample:
+                break
+            # check if the candidate has a vector
+            if candidate.vector is not None:
+                samples.append(candidate)
+
+        # can't build a matrix with less than 2 results
+        if len(samples) < 2:
+            return [], []
+
+        # sort samples by id
+        samples = sorted(samples, key=lambda x: x.id)
+        # extract the ids
+        ids = [sample.id for sample in samples]
+        scores: List[List[ScoredPoint]] = []
+
+        # Query `limit` neighbors for each sample
+        for sampled in samples:
+            sampled_id_index = ids.index(sampled.id)
```

Comment:
I used `ids` for `ids_to_includes = [x for (i, x) in enumerate(ids) if i != sampled_id_index]`

My idea was to avoid recomputing the set of the `ids` at every iteration just to remove one element.